### DISCUSSION TITLE
What are the precise step-by-step instructions for creating a cask in the Homebrew repo?

### DISCUSSION THREAD

**Discussion by shyamalschandra** [Homebrew/discussions#4640] [Casks]
### Output of `brew config`

```shell
n/a
```

### Output of `brew doctor`

```shell
n/a
```

### Description of issue

What are the precise step-by-step instructions for creating a cask in the Homebrew repo?  I have some proprietary app written for MacOS that I would like to distribute using Homebrew cask (e.g. https://formulae.brew.sh/cask/synthesia).  There is lots of unclarity in the instructions so please go through line-by-line on one of a few or more cask files so I can get the gist or rhythm of things on how to fill out the ruby script because there are so many options on the commands under brew.

### Relevant casks

Synthesia

**Comment by SMillerDev** [comment]
https://docs.brew.sh/Adding-Software-to-Homebrew

If this is not clear enough I don't think you'll be the best suited person to make the cask.

**Reply by johnpitchko** [reply]
One 'issue' I ran into is that, after following the steps in [How to Open a Homebrew Pull Request](https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request#cask-related-pull-request), the directions in [Adding Software to Homebrew](https://docs.brew.sh/Adding-Software-to-Homebrew) do not work 'out of the box'.

For example, running `brew create --cask download-url --set-name my-new-cask` creates the new cask in the Homebrew folder rather than the forked repo.

Then, even after creating the cask with proper stanza, running:

```
export HOMEBREW_NO_AUTO_UPDATE=1
brew install rippling
```
returns `Error: No formulae or casks found for rippling.`.

Now I certainly could be missing something, but I'm certain I'm following the directions as written.

**Reply by ZhongRuoyu** [reply]
> running `brew create --cask download-url --set-name my-new-cask` creates the new cask in the Homebrew folder rather than the forked repo.

Did you mean `$HOMEBREW_REPOSITORY/Library/Taps/homebrew/homebrew-cask`? That's expected, because that's where Homebrew works (e.g., it looks for casks there when not installing from API -- see below). Make sure to add your fork as a remote there.

> returns `Error: No formulae or casks found for rippling.`.

Please try

```shell
export HOMEBREW_NO_INSTALL_FROM_API=1
```

before doing `brew install`.

**Reply by SMillerDev** [reply]
> For example, running brew create --cask download-url --set-name my-new-cask creates the new cask in the Homebrew folder rather than the forked repo.

But where does it say to clone your forked repo in a local directory? Step 3 and 4 tell you how to add the fork as a remote to the Homebrew repository.

**Reply by johnpitchko** [reply]
Oops I misread the instructions in my eagerness to build the cask. My bad!

**Reply by johnpitchko** [reply]
That seemed to work, thank you @ZhongRuoyu . Should that `export` command be added to the docs? Happy to do that.

**Reply by ZhongRuoyu** [reply]
> Should that `export` command be added to the docs? Happy to do that.

Yes. The `HOMEBREW_NO_INSTALL_FROM_API` environment variable is needed since Homebrew 4.0.0, when API install became the default. Some parts of the docs are not updated accordingly, so please open a pull request. Thanks!

FYI, the docs are in [Homebrew/brew](https://github.com/Homebrew/brew), under the `docs` directory.

**Comment by apainintheneck** [comment]
I just reopened an issue that is tangentially related to messaging around `brew install` and local formula changes.

- https://github.com/Homebrew/brew/issues/15073
